soroban-spec/fixtures/ts/src/convert.ts

```diff
@@ -0,0 +1,129 @@
+        case undefined: {
+            return BigInt(0);
+        }
+        case xdr.ScValType.scvU64(): {
+            const { high, low } = scval.u64();
+            return bigIntFromBytes(false, [high, low]);
+        }
+        case xdr.ScValType.scvI64(): {
+            const { high, low } = scval.i64();
+            return bigIntFromBytes(true, [high, low]);
+        }
+        case xdr.ScValType.scvU128(): {
+            const parts = scval.u128();
+            const a = parts.hi();
+            const b = parts.lo();
+            return bigIntFromBytes(false, [a.high, a.low, b.high, b.low]);
+        }
+        case xdr.ScValType.scvI128(): {
+            const parts = scval.i128();
+            const a = parts.hi();
+            const b = parts.lo();
+            return bigIntFromBytes(true, [a.high, a.low, b.high, b.low]);
+        }
+        // case xdr.ScValType.scvU256(): {
+        //     return bigIntFromBytes(false, scval.u256());
+        // }
+        // case xdr.ScValType.scvI256(): {
+        //     return bigIntFromBytes(true, scval.i256());
+        // }
+        default: {
+            throw new Error(`Invalid type for scvalToBigInt: ${scval?.switch().name}`);
+        }
+    };
+}
+
+export function scValStrToJs<T>(base64_xdr: string): T {
+    let scval = xdr.ScVal.fromXDR(Buffer.from(base64_xdr, 'base64'));
+    return scValToJs(scval);
+}
+
+export function scValToJs<T>(val: xdr.ScVal): T {
+    switch (val?.switch()) {
+        case xdr.ScValType.scvBool():
+            return val.b() as T;
+        case undefined: {
+            return 0 as T;
+        }
+        case xdr.ScValType.scvU32(): {
+            return val.u32() as T;
+        }
+        case xdr.ScValType.scvI32(): {
+            return val.i32() as T;
+        }
+        case xdr.ScValType.scvU64():
+        case xdr.ScValType.scvI64():
+        case xdr.ScValType.scvU128():
+        case xdr.ScValType.scvI128():
+        case xdr.ScValType.scvU256():
+        case xdr.ScValType.scvI256():
+            return scvalToBigInt(val) as T;
+        case xdr.ScValType.scvAddress():
+            return Address.fromScVal(val).toString() as T;
+        case xdr.ScValType.scvString():
+            return val.str().toString() as T;
+        case xdr.ScValType.scvSymbol():
+            return val.sym().toString as T;
+        case xdr.ScValType.scvBytes():
+            return val.bytes() as T;
+        case xdr.ScValType.scvVec(): {
+            type Element = ElementType<T>;
+            return val.vec().map(scValToJs<Element>) as T
+        }
+        case xdr.ScValType.scvMap(): {
+            type Key = KeyType<T>;
+            type Value = ValueType<T>;
+            let res = new Map<Key, Value>();
+            val.map().forEach((e) => {
+                let key = scValToJs<Key>(e.key());
+                let value = scValToJs<Value>(e.val());
+                res.set(key as Key, value as Value);
+            });
+
+            return res as T
```

Comment:
TS is mad because without the casting it can't be sure that the specific type returned is `T`. e.g.
```ts
val.str().toString() as T
```
Here we either needed to assert that T is a string or cast it.

This is again an issue with doing the XDR -> JSON here. For one this we lack the ScTypeDef types which include the user defined types. So this type isn't even correct here because ScVal treats maps and structs the same.

Now that we are generating the exact package I think it would be worth returning to using rust for this parsing. Then we just need to ship the XDR.

This prevents having to have parity of JSON generators/parsers.

Reply:
eugh. that sucks. but fine for now.